HTTP headers: deprecating "(Want-)?Digest", adding "(Want-)?(Content-|Repr-)Digest" (#26885)
mdn/content: files/en-us/glossary/representation_header/index.md

page-type: glossary-definition
---

> A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes the particular _representation_ of the target resource sent in an HTTP message body.

Reviewer: Why "target"? 
My thinking here is that a representation header may be present in both request and response messages. So you can have a representation header without any resource in the message body. 

```suggestion
A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes the particular _representation_ of a resource.
```

Author: I also changed "the particular _representation_" to "one particular _representation_".

Representations are different forms of a particular resource.  
For example, the same data might be formatted as a particular media type such as XML or JSON, localized to a particular written language or geographical region, and/or compressed or otherwise encoded for transmission. The underlying resource is the same in each case, but its representation is different.  

> Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms: E.&hairsp;g. HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}} does *not* affect the representation, wheres e.&hairsp;g. {{HTTPHeader("Content-Range")}} does.  

Reviewer: ```suggestion
Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms.
For example, HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}} does *not* affect the representation, whereas e.g. {{HTTPHeader("Content-Range")}} does.  
```

Clients specify the formats that they prefer to be sent during [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation) (using `Accept-*` headers), and the representation headers tell the client the format of the _selected representation_ they received.

Representation headers may be present in both HTTP request and response messages.

> If sent as a response to a `HEAD` request, they describe the body content representation that _would_ be selected if the resource was requested with a `GET` request.

Reviewer: Given it says "representation headers may be present in both HTTP request and response messages, I wonder if we need  an example in a request?

Author: Both a positive and negative example may make sense: something akin to:
### positive example
```http
PUT /api/transact HTTP/1.1
Want-Content-Digest: sha-256
Content-Type: text/json
...
```
```http
HTTP/1.1 201 Created
Content-Digest: sha-256=:2IBI7hQn83oTCgB3Z/6apOl91WGoctRfRj/F9gkvVo8=:
Content-Encoding: br
Repr-Digest: sha-256=:W8oN3H3CmE/CBpV6ZPNozV2AIDzzQpWL7CCOXyDyDzI=:
...
```
### negative example
```http
GET /api/last-transaction HTTP/1.1
Accept: text/json, text/xml
Content-Digest: sha-256=:125+ms4/wPK0vMKM3oy9j4zu8HmPTmr7YVla229fhFc=:
...
```
```http
HTTP/1.1 406 Not Acceptable
Content-Digest: sha-256=:2IBI7hQn83oTCgB3Z/6apOl91WGoctRfRj/F9gkvVo8=:
...
```

Author: But I still wonder when it makes sense for the server to respond with `Want-Content-Digest`.

Reviewer: Yes, these kinds of examples make sense to me. And no, I don't understand when a server could ever response with Want-Content-Digest. I doubt its like those client hints where a server can specify a preference for future headers from the client.

If we can't prove it isn't a valid use case, I'd still just not cover this.

Author: Added in 72ee7b6a3b59f2f2ae41b848046c092390a48fd3.

> Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, {{HTTPHeader("Content-Location")}}, {{HTTPHeader("Content-Range")}}, {{Deprecated_Inline}}{{HTTPHeader("Digest")}} and {{HTTPHeader("Repr-Digest")}}.

Reviewer: ```suggestion
Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, {{HTTPHeader("Content-Location")}}, {{HTTPHeader("Content-Range")}}, {{HTTPHeader("Repr-Digest")}}, and {{HTTPHeader("Digest")}} {{Deprecated_Inline}}.
```

## See also

- [RFC 9110, section 3.2: Representations](https://httpwg.org/specs/rfc9110.html#representations)
- [List of all HTTP headers](/en-US/docs/Web/HTTP/Headers)
- {{Glossary("Payload header")}}
- {{glossary("Entity header")}}

> - {{Deprecated_Inline}}{{HTTPHeader("Digest")}}, {{Deprecated_Inline}}{{HTTPHeader("Want-Digest")}}
> - {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Repr-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}, {{HTTPHeader("Want-Repr-Digest")}}

Reviewer: Deprecated things are "less important" and should IMO go last.

```suggestion
- {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Repr-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}, {{HTTPHeader("Want-Repr-Digest")}}
- {{HTTPHeader("Digest")}} {{Deprecated_Inline}}, {{HTTPHeader("Want-Digest")}} {{Deprecated_Inline}}
```

page-type: glossary-definition
---

A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes one particular _representation_ of a resource.

Representations are different forms of a particular resource.

For example, the same data might be formatted as a particular media type such as XML or JSON<!--,--> or localized to a particular written language or geographical region<!--, and/or compressed or otherwise encoded for transmission-->. The underlying resource is semantically the same in each case (although details may vary: an e.&hairsp;g. text/html page is inherently more feature-rich than an e.&harisp;g. text/markdown page), but its representation is different.

> Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms: an HTTP message frames (cf. e.&hairsp;g. HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}) a particular stream of octets (cf. e.&hairsp;g. {{HTTPHeader("Content-Range")}}) derived from the _selected representation_.

Reviewer: Can you globally purge the use of `&hairsp;`?

We might add a comment here about the fact that in some case `<!--, and/or compressed or otherwise encoded for transmission-->` is relevant?

Reviewer: P.S. It doesn't quite "fit", but could perhaps be a "Note" since it is related, but also an aside.

Clients specify the formats that they prefer to be sent during [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation) (using `Accept-*` headers), and the representation headers tell the client the format of the _selected representation_ they received.

Representation headers may be present in both HTTP request and response messages with various methods.
If sent as a response to a `HEAD` request, they describe the body content representation that _would_ be selected if the resource was requested with a `GET` request.

Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, {{HTTPHeader("Content-Range")}}, {{HTTPHeader("Repr-Digest")}}, and {{HTTPHeader("Digest")}} {{Deprecated_Inline}}.

> ## Examples

Reviewer: Actually, I was wrong sorry. We try and avoid having examples in glossary docs. We try to keep the glossary as short as possible. These examples should exist, if they do not, in the reference pages - e.g. want-repr-digest.

Author: It didn't fit in the glossary. See bc31ed5466175c2e3f10fb759779cb43ddfecca2.

{{GlossarySidebar}}

A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes one particular _representation_ of a resource.

Representations are different forms of a particular resource.

For example, the same data might be formatted as a particular media type such as XML, JSON, HTML or Markdown, or localized to a particular written language or geographical region.
The underlying resource is semantically the same in each case, but its representation is different.

> Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms: an HTTP message frames (cf. e.&hairsp;g. HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}) a particular stream of octets (cf. e.&hairsp;g. {{HTTPHeader("Content-Range")}}) derived from the _selected representation_.

Reviewer 2: ```suggestion
While representations are different forms of resources, representations can themselves also be transmitted in various forms: an HTTP message frames (cf., e.g., HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}), a particular stream of octets (cf., e.g., {{HTTPHeader("Content-Range")}}) derived from the _selected representation_.
```

Clients specify the formats that they prefer to be sent during [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation) (using `Accept-*` headers), and the representation headers tell the client the format of the _selected representation_ they received.

Representation headers may be present in both HTTP request and response messages with various methods.
If sent as a response to a `HEAD` request, they describe the body content representation that _would_ be selected if the resource was requested with a `GET` request.

Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, {{HTTPHeader("Content-Range")}}, {{HTTPHeader("Repr-Digest")}}, and {{HTTPHeader("Digest")}} {{Deprecated_Inline}}.

## See also

- [RFC 9110, section 3.2: Representations](https://httpwg.org/specs/rfc9110.html#representations)
- [List of all HTTP headers](/en-US/docs/Web/HTTP/Headers)
- {{Glossary("Payload header")}}
- {{glossary("Entity header")}}
- {{HTTPHeader("Repr-Digest")}}, {{HTTPHeader("Want-Repr-Digest")}}

> - pertaining to particular forms of the selected representation: {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}

Reviewer 2: ```suggestion
- {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}
```

A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes one particular _representation_ of a resource.

Representations are different forms of a particular resource.  

> For example, the same data might be formatted as a particular media type such as XML or JSON<!--,--> or localized to a particular written language or geographical region<!--, and/or compressed or otherwise encoded for transmission-->. The underlying resource is semantically the same in each case (although details may vary: an e.&hairsp;g. text/html page is inherently more feature-rich than an e.&harisp;g. text/markdown page), but its representation is different.

Author: @hamishwillee This is a tad tricky: the deprecated "Digest" header depended on e.&hairsp;g. "Content-Encoding". But "selected representation" in the "Repr-Digest"-sense does not depend on "Content-Encoding"; "Content-Digest", however, does.

Author: Ok.
